proto/asn1-pdu-proto/asn1_proto_to_der.h

```diff
@@ -0,0 +1,74 @@
+#ifndef ASN1_PROTO_TO_DER_H_
+#define ASN1_PROTO_TO_DER_H_
+
+#include <cmath>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <typeinfo>
+#include <vector>
+#include "asn1_pdu.pb.h"
+
+namespace asn1_pdu {
+
+class ASN1ProtoToDER {
+ public:
+  // Encodes |pdu| to DER, returning the encoded bytes of the PDU in
+  // |encoder_|.
+  std::vector<uint8_t> ProtoToDER(const PDU& pdu);
+
+ private:
+  std::vector<uint8_t> encoder_;
+
+  // Enocdes |pdu| to DER, returning the number of bytes used to encode |pdu|.
+  void EncodePDU(const PDU& pdu);
+
+  // Encodes |id| to DER according to X.690 (2015), 8.1.2.
+  // Returns number of bytes used to encode |id|.
+  void EncodeIdentifier(const Identifier& id);
+
+  // Concatinates |id_class|, |encoding|, and |tag| according to DER
+  // high-tag-number form rules (X.690 (2015), 8.1.2.4).
+  void EncodeHighTagNumberForm(const uint8_t id_class,
+                               const uint8_t encoding,
+                               const uint32_t tag);
+
+  // Encodes the length to DER.
+  // |len| can be used to affect the encoding, in order to produce
+  // invalid lengths. |actual_len| is the correct length of the PDU,
+  // and is used when |len| is not. |len_pos| contains the offset in |encoder_|
+  // where the length should be encoded.
+  // To correctly call this, the tag must already be encoded immediately prior
+  // to |len_pos|, and the remainder of |encoder_| represents the encoded value.
+  void EncodeLength(const Length& len, size_t actual_len, size_t len_pos);
+
+  // Writes |raw_len| to |encoder_| at |len_pos|.
+  void EncodeOverrideLength(const std::string raw_len, const size_t len_pos);
```

Comment:
```suggestion
  void EncodeOverrideLength(const std::string& raw_len, const size_t len_pos);
```

And update the corresponding implementation. This avoids forcing a copy of raw_len to be made